### ISSUE TITLE
[Bug]  SealedSecret error with the MariaDb random password generation

### ISSUE THREAD

**Issue by K-MeLeOn** [mariadb-operator/mariadb-operator#340]
**Documentation**
- [x] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**

When using SealedSecret, MariaDb Operator generate a random password too fast for him. It can't take the owner lead over MariaDb Operator.

**Expected behaviour**

Do not generate a random password or wait a bit before create one.

**Steps to reproduce the bug**

1. Install SealedSecret
2. Create a SealedSecret for the root and the user
3. Publish an app using MariaDb with the SealedSecret
4. Observe the lead taken by MariaDb over SealedSecret to unseal and generate the k8s secret

**Debug information**

See the pictures:
![20240114_130600](https://github.com/mariadb-operator/mariadb-operator/assets/47109782/7aba91a9-9251-45a3-9256-41da3eb17826)
![20240114_131440](https://github.com/mariadb-operator/mariadb-operator/assets/47109782/b9a74f8a-13f2-430d-a8ba-594c0213193e)

NOK secret (owner is MariaDb):
```YAML
apiVersion: v1
kind: Secret
metadata:
  name: db-su-creds
  namespace: ++++++++
  uid: 8c88d378-9c8d-4ef8-9ad2-a136d7f589ff
  resourceVersion: '7254051'
  creationTimestamp: '2024-01-14T11:52:12Z'
  ownerReferences:
    - apiVersion: mariadb.mmontes.io/v1alpha1
      kind: MariaDB
      name: ++++++++
      uid: 105e5bf6-3d11-4861-9bc2-acf054297d74
      controller: true
      blockOwnerDeletion: true
data:
  root-password: ++++++++
type: Opaque
```

OK secret (owner is SealedSecret):
```YAML
apiVersion: v1
data:
  password: ++++++++
  username: ++++++++
kind: Secret
metadata:
  creationTimestamp: '2024-01-14T11:43:55Z'
  name: db-u-creds
  namespace: ++++++++
  ownerReferences:
    - apiVersion: bitnami.com/v1alpha1
      controller: true
      kind: SealedSecret
      name: db-u-creds
      uid: 6e4d4d1a-697a-45a7-9813-aff65f7698d7
  resourceVersion: '7244775'
  uid: 9925b3d4-3f0d-413b-8de8-7254c1483d35
type: Opaque
```

**Environment details**:
- Kubernetes version: 1.27.9
- Kubernetes distribution: k3s
- mariadb-operator version: 0.24
- Install method: helm w/ ArgoCd
- Install flavor: recommended full

**Comment by K-MeLeOn** [comment]
For the time it can be bypassed by generating SealedSecret secrets before adding MariaDb, but this breaks the main function of GitOps w/ ArgoCD (and Flux ?) for automated deployment.

**Comment by mmontes11** [comment]
Hey there @K-MeLeOn ! Thanks for reporting.

This indeed could happen, as our `Secret` reconciler does generate a random secret if the referred `Secret` is not found, see:
https://github.com/mariadb-operator/mariadb-operator/blob/03a28b35b920bc33b0b59b0bee0152e00cc584b8/pkg/controller/secret/controller.go#L27

Since we are reusing this controller for multiple resources, I suggest we add a flag `--generate-random-passwords` to the controller cmd to explicitly declare that we want the random secret to be generated as they do currently. This would be the default behaviour but we could potentially disable this by passing `--generate-random-passwords=false`. In the last case, the controller will fail if the `Secret` is not present, causing the reconcile request to be requeued until we eventually have a `Secret` i.e. the `SealedSecret` has been reconciled.

Thoughts? 

As an alternative we can do this on a per resource basis i.e. add a `spec.generateRandomPasswords` to every CRD that has password references, but I lean more towards the flag option.

**Comment by melledouwsma** [comment]
> I suggest we add a flag `--generate-random-passwords` to the controller cmd to explicitly declare that we want the random secret to be generated as they do currently. 

That a nice addition. One could also use Argo CD [Sync waves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) or Flux [dependsOn](https://fluxcd.io/flux/components/kustomize/kustomizations/#dependencies) to make sure the `SealedSecret` is deployed before the `MariaDB` instance.

**Comment by mmontes11** [comment]
> That a nice addition. One could also use Argo CD [Sync waves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) or Flux [dependsOn](https://fluxcd.io/flux/components/kustomize/kustomizations/#dependencies) to make sure the SealedSecret is deployed before the MariaDB instance.

Thanks for your input @melledouwsma . I'm more into flux, and can confirm that the `dependsOn` will work.

**Comment by K-MeLeOn** [comment]
> > I suggest we add a flag `--generate-random-passwords` to the controller cmd to explicitly declare that we want the random secret to be generated as they do currently.
> 
> That a nice addition. One could also use Argo CD [Sync waves](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/) or Flux [dependsOn](https://fluxcd.io/flux/components/kustomize/kustomizations/#dependencies) to make sure the `SealedSecret` is deployed before the `MariaDB` instance.

Oh thank you, I completely forgot that ! I'll add some waves now :+1:



> Hey there @K-MeLeOn ! Thanks for reporting.
> 
> This indeed could happen, as our `Secret` reconciler does generate a random secret if the referred `Secret` is not found, see:
> 
> https://github.com/mariadb-operator/mariadb-operator/blob/03a28b35b920bc33b0b59b0bee0152e00cc584b8/pkg/controller/secret/controller.go#L27
> 
> Since we are reusing this controller for multiple resources, I suggest we add a flag `--generate-random-passwords` to the controller cmd to explicitly declare that we want the random secret to be generated as they do currently. This would be the default behaviour but we could potentially disable this by passing `--generate-random-passwords=false`. In the last case, the controller will fail if the `Secret` is not present, causing the reconcile request to be requeued until we eventually have a `Secret` i.e. the `SealedSecret` has been reconciled.
> 
> Thoughts?
> 
> As an alternative we can do this on a per resource basis i.e. add a `spec.generateRandomPasswords` to every CRD that has password references, but I lean more towards the flag option.

Uuh, I'm a rookie in project structure, does the flag stop password generation for all MariaDb instances or only for the flagged instance ?

**Comment by Hoega** [comment]
> Uuh, I'm a rookie in project structure, does the flag stop password generation for all MariaDb instances or only for the flagged instance ?

Depending on the implementation, the flag can apply to all instances if specified at the controller deployment level. Alternatively, if an option is added at the CRD level, each instance can have its flag.

**Comment by mmontes11** [comment]
This has been addressed by https://github.com/mariadb-operator/mariadb-operator/pull/598 and released in [v0.0.28](https://github.com/mariadb-operator/mariadb-operator/releases/tag/v0.0.28). 

See also a working example with sealed-secrets + flux:
- https://github.com/mariadb-operator/mariadb-operator/tree/main/examples/flux/infrastructure/mariadb

Closing!
